### ISSUE TITLE
Monomorphize public API

### ISSUE THREAD

**Issue by Ralith** [quinn-rs/quinn#911]
Quinn's public API is complicated by abstraction over the cryptographic backend, which is of niche interest. While the crypto traits themselves have good reasons not to be object-safe, we should be able to encapsulate the surrounding logic into an internal object-safe interface, and thereby erase the type parameter that's spread everywhere. By avoiding a need for downstream monomorphization, this may also significantly improve downstream compile times.

**Comment by burdges** [comment]
I'd think binaries should never have more than one cryptographic backend, which usually favors monomorphization over trait objects.

**Comment by Ralith** [comment]
My expectation is that the cost of a trait object will be a single-digit number of virtual calls per packet, which should be insignificant. We'll of course sanity-check against the bulk throughput benchmark before committing to a new approach.

**Comment by DemiMarie** [comment]
Even a single-digit number of calls per packet can be significant at high data rates.  That said, someone flooding a QUIC implementation with lots of tiny packets is probably either trying to perform a DoS attack or has a buggy implementation.

**Comment by infinity0** [comment]
Do you have a concrete idea for doing this? I don't think this is avoidable if you want users to be able to use different crypto implementations. Fundamentally, the user will need to be able to choose their own crypto key type, and some part of the public API will have to expose the concrete key type that the user chose (rather than just a trait), e.g. to get the peer's public key for a given connection.

What you can do however is have a mirror API that wraps everything in a 1-field struct that hard-codes `TlsSession`, then the monomorphisation will be down when compiling quinn rather than downstream.

Also, one parameter is the maximum you need - everything else can be associated types on (some common trait implemented by) that concrete type. This could include things like `dyn ConnectionIdGenerator` and `dyn Controller` that I pointed out in #927, avoiding `dyn` and making type-safe serialisation easier to implement.

**Comment by Ralith** [comment]
> some part of the public API will have to expose the concrete key type that the user chose

No such API currently exists, unless I'm confused about your meaning. The choice of key can be isolated within the type erasure boundary.

**Comment by infinity0** [comment]
> No such API currently exists, unless I'm confused about your meaning.

The current API has a parameter for the security session as a trait, which has various associated types including the public key type (AuthenticationData). I assumed there was an actual method you could call, it turns out not, but you could easily add one without changing anything else, as long as the concrete session type S is exposed as it is currently.

If quinn erases the type in the API, this concrete type is lost and the user wouldn't be able to recover out of quinn, since it is the responsibility of the quinn logic to instantiate each new session, so the user never has a reference to the concrete type of it. (The user has a reference to the concrete ClientConfig<S> with a concrete S, however they will never see an actual concrete S reference, and therefore cannot recover the concrete public key type.) There are various ways around this such as allowing the user to pass in a callback where quinn calls it with the concrete S every time it creates one, but this is a bit hacky.

**Comment by infinity0** [comment]
> The choice of key can be isolated within the type erasure boundary.

Yeah, I'm curious as my current understanding based on the above reasoning is it's not really feasible in a clean way, did you have some other ideas?

**Comment by Ralith** [comment]
I'm not sure I understand your question. A purely hypothetical API which we don't plan to add is not a blocker.

**Comment by infinity0** [comment]
Getting the peer public key from a secure connection, is a common API that any secure networking or messaging protocol should be supporting. Otherwise it's harder to build higher-level protocols on top of it. Why don't you plan to add it?

**Comment by infinity0** [comment]
My point is also more general - crypto sessions can be customised, and the user who selects it would often like to take concrete types out of it, but quinn is the one responsible for instantiating sessions per stream (which is clean & fine IMO), leading to the issue I observed above.

**Comment by Ralith** [comment]
We already provide an API to access the crypto state itself, which more specialized accessors can be built on. That can be preserved under monomorphization via downcasting

**Comment by infinity0** [comment]
Downcasting is another way around it for sure, but it uses dynamic typing and carries a runtime cost. It also seems there's no standard way to do it in rust, I can see [downcast](https://crates.io/crates/downcast) and [downcast-rs](https://crates.io/crates/downcast-rs) on a quick search. From reading their docs, it seems it is the trait that has to opt into it (i.e. quinn for Session here), presumably to provide the necessary runtime fields for checking the type.

**Comment by Ralith** [comment]
Drafted at https://github.com/quinn-rs/quinn/pull/1201.
